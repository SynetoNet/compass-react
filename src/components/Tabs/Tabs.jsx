import React from "react";
import PropTypes from "prop-types";

import BootstrapTabs from "react-bootstrap/Tabs";
import BootstrapTab from "react-bootstrap/Tab";
import "./Tabs.scss";

class Tabs extends React.Component {
  static Tab = BootstrapTab;
  constructor(props) {
    super(props);
  }
  render() {
    return (
      <div
        className={`tabs-${this.props.layout} alignment-${
          this.props.alignment
        }`}
      >
        <BootstrapTabs {...this.props} />
      </div>
    );
  }
}

Tabs.propTypes = {
  layout: PropTypes.oneOf(["vertical", "horizontal"]),
  alignment: PropTypes.oneOf(["left", "right"])
};

Tabs.defaultProps = {
  layout: "horizontal",
  alignment: "left"
};

export default Tabs;
