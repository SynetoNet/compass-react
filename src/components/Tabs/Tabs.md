### **Description**

Used when a lot of information should be displayed in a small amount of space.
Information is displayed in tabs by grouping it in several categories, each category becoming a tab head.

<br />
<br />

#### **Default Tabs**

By default, tabs are horizontal and left aligned:

```jsx
<Tabs>
  <Tabs.Tab eventKey="home" title="Home">
    My heavy fox, you inspire me to write. How I love the way you skip, sleep
    and sneer, Invading my mind day and through the night, Always dreaming about
    the cunning smear.
  </Tabs.Tab>
  <Tabs.Tab eventKey="profile" title="Profile">
    Let me compare you to an ideal spoon? You are more deadpan, stunning and
    scaly. Fun sun heats the surreal peaches of June, And summertime has the
    vernal qualey.
  </Tabs.Tab>
  <Tabs.Tab eventKey="contact" title="Contact" disabled>
    How do I love you? Let me count the ways. I love your clad ankle, elbows and
    smile. Thinking of your glad elbows fills my days. My love for you is the
    bally air mile.
  </Tabs.Tab>
</Tabs>
```

#### **Right aligned**

```jsx
<Tabs align="end">
  <Tabs.Tab eventKey="home" title="Home">
    My heavy fox, you inspire me to write. How I love the way you skip, sleep
    and sneer, Invading my mind day and through the night, Always dreaming about
    the cunning smear.
  </Tabs.Tab>
  <Tabs.Tab eventKey="profile" title="Profile">
    Let me compare you to an ideal spoon? You are more deadpan, stunning and
    scaly. Fun sun heats the surreal peaches of June, And summertime has the
    vernal qualey.
  </Tabs.Tab>
  <Tabs.Tab eventKey="contact" title="Contact" disabled>
    How do I love you? Let me count the ways. I love your clad ankle, elbows and
    smile. Thinking of your glad elbows fills my days. My love for you is the
    bally air mile.
  </Tabs.Tab>
</Tabs>
```

#### **Vertical Tabs**

```jsx
<Tabs layout="vertical">
  <Tabs.Tab eventKey="home" title="Home">
    My heavy fox, you inspire me to write. How I love the way you skip, sleep
    and sneer, Invading my mind day and through the night, Always dreaming about
    the cunning smear.
  </Tabs.Tab>
  <Tabs.Tab eventKey="profile" title="Profile">
    Let me compare you to an ideal spoon? You are more deadpan, stunning and
    scaly. Fun sun heats the surreal peaches of June, And summertime has the
    vernal qualey.
  </Tabs.Tab>
  <Tabs.Tab eventKey="contact" title="Contact" disabled>
    How do I love you? Let me count the ways. I love your clad ankle, elbows and
    smile. Thinking of your glad elbows fills my days. My love for you is the
    bally air mile.
  </Tabs.Tab>
</Tabs>
```

#### **Custom title**

The **`title`** prop supports any React Component:

```jsx
import Badge from "../Badge/Badge";

const Inbox = (
  <>
    Inbox{" "}
    <Badge color="success" variant="strong" width={32} className="ml-2">
      2
    </Badge>
  </>
);

const Urgent = (
  <>
    Urgent{" "}
    <Badge color="danger" variant="strong" width={32} className="ml-2">
      2
    </Badge>
  </>
);

const Trash = (
  <>
    Trash{" "}
    <Badge variant="outline" width={32}>
      99+
    </Badge>
  </>
);

<Tabs>
  <Tabs.Tab eventKey="home" title={Inbox}>
    My heavy fox, you inspire me to write. How I love the way you skip, sleep
    and sneer, Invading my mind day and through the night, Always dreaming about
    the cunning smear.
  </Tabs.Tab>
  <Tabs.Tab eventKey="profile" title={Urgent}>
    Let me compare you to an ideal spoon? You are more deadpan, stunning and
    scaly. Fun sun heats the surreal peaches of June, And summertime has the
    vernal qualey.
  </Tabs.Tab>
  <Tabs.Tab eventKey="contact" title={Trash} disabled>
    How do I love you? Let me count the ways. I love your clad ankle, elbows and
    smile. Thinking of your glad elbows fills my days. My love for you is the
    bally air mile.
  </Tabs.Tab>
</Tabs>;
```

#### **Extra content**

You can render any additional content inline with the Tab Navigation. Your content will be rendered inside a flex container, that fills the entire empty space in the Navigation.

```jsx
import Button from "../Button/Button";

<Tabs
  extra={
    <>
      <strong style={{ flex: 1 }}>Extra title</strong>
      <Button variant="link">Extra link</Button>
    </>
  }
  align="end"
>
  <Tabs.Tab eventKey="home" title="Home">
    My heavy fox, you inspire me to write. How I love the way you skip, sleep
    and sneer, Invading my mind day and through the night, Always dreaming about
    the cunning smear.
  </Tabs.Tab>
  <Tabs.Tab eventKey="profile" title="Profile">
    Let me compare you to an ideal spoon? You are more deadpan, stunning and
    scaly. Fun sun heats the surreal peaches of June, And summertime has the
    vernal qualey.
  </Tabs.Tab>
</Tabs>;
```

#### **Controlled Tabs**

By default, the Tabs are uncontrolled, having their own internal state regarding the activeTab. If you specify the **`activeKey`** and the **`onSelect`** handler, you can turn this into a controlled component:

```jsx
import { useState } from "react";

const ControlledTabs = () => {
  const [activeKey, setActiveKey] = useState("home");

  return (
    <>
      <Tabs activeKey={activeKey} onSelect={tab => setActiveKey(tab)}>
        <Tabs.Tab eventKey="home" title="Home">
          Home tab content, my heavy fox, you inspire me to write.
        </Tabs.Tab>
        <Tabs.Tab eventKey="profile" title="Profile">
          Profile tab, let me compare you to an ideal spoon?
        </Tabs.Tab>
      </Tabs>
    </>
  );
};

<ControlledTabs />;
```

#### **Extra Dropdown**

To handle additional actions, it is required to make the Tab a controlled component:

```jsx
import { useState } from "react";
import Dropdown from "../Dropdown/Dropdown";

const ControlledTabsWithDropdown = () => {
  const [activeKey, setActiveKey] = useState("home");

  return (
    <>
      <Tabs
        activeKey={activeKey}
        onSelect={tab => {
          setActiveKey(tab);
          console.log("selected tab", tab);
        }}
        extra={
          <Dropdown
            label="Dropdown"
            variant="tertiary"
            onSelect={eventKey =>
              console.log("selected dropdown option", eventKey)
            }
            items={[{ label: "Option 1" }, { label: "Option 2" }]}
          />
        }
      >
        <Tabs.Tab eventKey="home" title="Home">
          Home tab content, my heavy fox, you inspire me to write.
        </Tabs.Tab>
        <Tabs.Tab eventKey="profile" title="Profile">
          Profile tab, let me compare you to an ideal spoon?
        </Tabs.Tab>
      </Tabs>
    </>
  );
};

<ControlledTabsWithDropdown />;
```
