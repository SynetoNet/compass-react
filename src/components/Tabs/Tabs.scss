@import "../../base/colors";
@import "../../base/variables";

// override Bootstrap styles
.nav-tabs {
  .nav-link {
    color: $neutral-100;
    border-top-left-radius: 0;
    border-top-right-radius: 0;
    position: relative;

    &.active {
      color: $accent;
      border-color: #dde3f0 #dde3f0 #fff;
    }

    &:hover::before,
    &.active::before {
      display: block;
      background: $accent;
      content: "";
      position: absolute;
    }
  }
}

// add Compass specific styles for Horizontal tabs
.tabs-horizontal {
  .tab-content {
    margin-top: 20px;
  }

  .nav-link {
    // remove hover border
    &:hover:not(.active) {
      border-color: transparent;
      color: $accent;
    }

    &.active::before {
      height: 3px;
      top: -1px;
      left: -1px;
      right: -1px;
    }
  }

  .nav-item {
    // when we have extra content, with higher height
    // we need to align items to bottom
    text-align: center;
    display: flex;
    &.tabs-extra {
      align-self: stretch;
    }
    .nav-link.active {
      height: 100%;
    }
  }

  &.tabs-align-end .nav-tabs {
    justify-content: flex-end;
  }
}

// add Compass specific styles for Vertical tabs
.tabs-vertical {
  display: flex;
  border: none;

  .tab-content {
    margin-left: 20px;
  }

  .nav-tabs {
    flex-flow: column;
    border-bottom: none;
    border-right: 1px solid $neutral-12;

    .nav-link {
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
      max-width: 200px;
      // remove hover border
      &:hover:not(.active) {
        border-color: transparent;
        color: $accent;
      }
      &.active {
        border-right: none;
        border-bottom: 1px solid $neutral-12;
        border-left-color: transparent;
        margin-right: -1px;
      }

      &.active::before {
        width: 3px;
        top: -1px;
        left: 0;
        bottom: -1px;
      }
    }
  }
}

.tabs-extra {
  flex: 1;
  align-self: stretch;
  display: flex;
  align-items: center;
  // this is to accomodate the top border of tab buttons
  // so content is vertically centered
  padding-top: 1px;
}
