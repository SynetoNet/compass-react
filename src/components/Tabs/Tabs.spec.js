import React from "react";
import Tabs from "./Tabs";
import renderer from "react-test-renderer";

test("renders Horizontal Tabs", () => {
  const component = renderer.create(
    <Tabs>
      <Tabs.Tab eventKey="tab1" title="tab1 title">
        tab1 content
      </Tabs.Tab>
      <Tabs.Tab eventKey="tab2" title="tab2 title">
        tab2 content
      </Tabs.Tab>
      <Tabs.Tab eventKey="tab3" title="tab3 title">
        tab3 content
      </Tabs.Tab>
    </Tabs>
  );
  expect(component.toJSON()).toMatchSnapshot();
});

test("renders Vertical Tabs", () => {
  const component = renderer.create(
    <Tabs vertical>
      <Tabs.Tab eventKey="tab1" title="tab1 title">
        tab1 content
      </Tabs.Tab>
      <Tabs.Tab eventKey="tab2" title="tab2 title">
        tab2 content
      </Tabs.Tab>
      <Tabs.Tab eventKey="tab3" title="tab3 title">
        tab3 content
      </Tabs.Tab>
    </Tabs>
  );
  expect(component.toJSON()).toMatchSnapshot();
});
