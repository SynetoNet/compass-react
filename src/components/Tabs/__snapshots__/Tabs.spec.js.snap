// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Horizontal Tabs 1`] = `
<div
  className="tabs-horizontal"
>
  <nav
    className="nav nav-tabs"
    layout="horizontal"
    onKeyDown={[Function]}
    role="tablist"
  >
    <a
      aria-controls={null}
      aria-selected={true}
      className="nav-item nav-link active"
      data-rb-event-key="tab1"
      href="#"
      id={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
    >
      tab1 title
    </a>
    <a
      aria-controls={null}
      aria-selected={false}
      className="nav-item nav-link"
      data-rb-event-key="tab2"
      href="#"
      id={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={-1}
    >
      tab2 title
    </a>
    <a
      aria-controls={null}
      aria-selected={false}
      className="nav-item nav-link"
      data-rb-event-key="tab3"
      href="#"
      id={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={-1}
    >
      tab3 title
    </a>
  </nav>
  <div
    className="tab-content"
  >
    <div
      aria-hidden={false}
      aria-labelledby={null}
      className="fade tab-pane active show"
      id={null}
      role="tabpanel"
    >
      tab1 content
    </div>
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    >
      tab2 content
    </div>
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    >
      tab3 content
    </div>
  </div>
</div>
`;

exports[`renders Vertical Tabs 1`] = `
<div
  className="tabs-horizontal"
>
  <nav
    className="nav nav-tabs"
    layout="horizontal"
    onKeyDown={[Function]}
    role="tablist"
    vertical={true}
  >
    <a
      aria-controls={null}
      aria-selected={true}
      className="nav-item nav-link active"
      data-rb-event-key="tab1"
      href="#"
      id={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
    >
      tab1 title
    </a>
    <a
      aria-controls={null}
      aria-selected={false}
      className="nav-item nav-link"
      data-rb-event-key="tab2"
      href="#"
      id={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={-1}
    >
      tab2 title
    </a>
    <a
      aria-controls={null}
      aria-selected={false}
      className="nav-item nav-link"
      data-rb-event-key="tab3"
      href="#"
      id={null}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      tabIndex={-1}
    >
      tab3 title
    </a>
  </nav>
  <div
    className="tab-content"
  >
    <div
      aria-hidden={false}
      aria-labelledby={null}
      className="fade tab-pane active show"
      id={null}
      role="tabpanel"
    >
      tab1 content
    </div>
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    >
      tab2 content
    </div>
    <div
      aria-hidden={true}
      aria-labelledby={null}
      className="fade tab-pane"
      id={null}
      role="tabpanel"
    >
      tab3 content
    </div>
  </div>
</div>
`;
